### DISCUSSION TITLE
Prisma client connection pool issues, timeouts when trying to use a new connection but DB pooler is never under load

### DISCUSSION THREAD

**Discussion by MorenoMdz** [prisma/prisma#24846] [Q&A]
### Question

Hey there, we have a very weird behavior in our system right now. We used to have problems with our services losing DB connectivity and DB connection count being too high in the past, which we mainly solved by rebuilding our infrastructure in a way that the DB servers and our API services are in the same region, and running thru a VPC.

It was working beautifully for weeks until recently we started getting Prisma pool timeout errors again. We have tried to identify potential issues at our application that could have caused this change but so far we haven't found a pattern that could indicate we have an issue in our application.

Our current configuration is:

* PG running in GPC (Aiven is our provider), mounted on a VPC
* pgbouncer set to 667 max connections and pg set to max 800 connections
* NodeJs with Prisma 5.4.2 running in GCP on the same VPC
* Our Node services have up to a max of 15 instances, with the average of 8 instances running
* Each instance has the Prisma client pool set to max 55 connections
* Our API, considering all available instances has an average of 8 requests/s with spikes of 100 requests/s, and is very Read heavy, I would say about 95/5 r/w.

We have been optimizing our queries the best we can, we do have a few sub optimal queries that are hard to index, and so far we haven't been able to associate the connection pool timeout issues with any of the known slow queries. We have been monitoring our queries performance very closely.

The metrics we get from Aiven to DataDog indicate the DB is never running out of connections, in this graph you can see the improvements in connection count from the PG side:

<img width="1732" alt="image" src="https://github.com/user-attachments/assets/1fd1b7f4-4148-4312-8a66-4b5c88e9e58c">

The metrics from PG we have in DD show the count of **active** connections; while we did suspect these numbers to be too low, it does match what we see in the Aiven dashboard under active queries. At the same time we always have lots of idle connections in Aiven:

<img width="1218" alt="image" src="https://github.com/user-attachments/assets/c8f67bbe-5bb6-4c60-9400-6e4f4415815d">

The PG server is rarely going above 20% CPU, memory or i/o load, so I don't think we have an issue with the configuration size. It doesn't look like the service is getting resource-constrained at any time.

Now that we have the scenario explained, we have a few questions:

1. How does the Prisma db pool work, does it starts the connection before processing the Prisma client query into the final SQL query/statement or does it build the query first then it tries to connect?

We are wondering if the 55 connections number could be a problem causing the Prisma engine to not be able to build some queries efficiently causing the timeouts. For context, a week ago we had an issue of a Prisma query causing our services to spike to 100% CPU which we fixed by using RAW SQL, I can make a ticket with that problematic query too if you want.

2. Increasing the Prisma pool timeout here might help or make it worse?

3. Is there a way for us to identify which queries are happening when the timeouts happened? Does Prisma expose any tools that could he use understand which queries were active in the client pool at a given time?

4. In our APM profiler, Prisma represents around 20%+ of the Node wall time per minute, is this an expected overhead for the ORM or could it indicate we have specific queries that are underperforming causing stress in the builder?
<img width="1293" alt="image" src="https://github.com/user-attachments/assets/f6d2fa6b-b61d-4d46-b5cc-e163bb6025ec">

Any help is appreciated here, I don't want to start tweaking timeouts and connection limits without understanding better what could be causing this.

Once again thank you very much for all the work done with Prisma and the help provided here!

### Information about Prisma Schema, Client Queries and Environment (optional)

I think this is the only relevant part of our schema.
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["postgresqlExtensions", "metrics", "views"]
}

datasource db {
  provider   = "postgresql"
  url        = env("DATABASE_URL")
  extensions = [postgis()]
}
```

- Database: PG
- Node.js version: 20.11
- Prisma 5.4.2

**Comment by ludralph** [comment]
Hi @MorenoMdz 👋 

Thank you for the detailed explanation of your setup and current issue. 

>1.  How does the Prisma db pool work, does it starts the connection before processing the Prisma client query into the final SQL query/statement or does it build the query first then it tries to connect?

Prisma connection pool works by:

1. Instantiating a connection pool with a configurable pool size and pool timeout.
2. Creating one initial connection to the pool.
3. Reserving a connection from the pool when a query comes in.
4. Opening additional connections if no idle connections are available, up to the connection limit.
5. Adding queries to a FIFO queue if no connections are available and throwing a timeout error (P2024) if queries wait too long.
 
Thus, Prisma attempts to reserve a connection from the pool before processing the query. The connection pool configuration plays a crucial role in query execution efficiency. If the pool size or timeout is insufficient, it can lead to timeouts. See this [section](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/connection-pool#how-the-connection-pool-works) of the documentation.

> 2. Increasing the Prisma pool timeout here might help or make it worse?

Increasing the pool timeout allows queries to wait longer for an available connection, potentially reducing timeout errors but may also increase overall response time. I would recommend 

- Increase that pool timeout value beyond 10s. Does this get rid of the errors at the cost of longer response times? Then:
- Increase the number of connections you give your applications, and increase the pool size by increasing the connection_limit value.

>3. Is there a way for us to identify which queries are happening when the timeouts happened? Does Prisma expose any tools that could be used to understand which queries were active in the client pool at a given time?

Prisma Client [Metrics](https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/metrics) give you a detailed insight into how Prisma Client interacts with your database. You can use this insight to help diagnose performance issues with your application. Enable and monitor Prisma metrics to gain insights into query performance and connection pool usage.

>4. In our APM profiler, Prisma represents around 20%+ of the Node wall time per minute, is this an expected overhead for the ORM or could it indicate we have specific queries that are underperforming causing stress in the builder?

Prisma introduces some overhead due to query building, connection management, and other abstractions. A 20%+ overhead could indicate that specific queries are underperforming and causing stress You can use tools like [EXPLAIN ANALYZE](https://www.postgresql.org/docs/current/sql-explain.html) to analyze the performance of your queries and identify any missing indexes or other issues.

**Reply by MorenoMdz** [reply]
So there is a chance some of the more complex queries we have could be causing this, saturating the client side pool, but not translate to a high connection count in pgbouncer/db, is this correct?

Some of our more complex queries involve deep nested select which sometimes require some extra filtering at the nested select level, do you think upgrading Prisma and trying the new join settings could help with this?

Any suggestions on how to identify potentially problematic queries other than adding prisma metrics to all suspect queries?

Also, I am assuming prisma Raw bypasses mostly the query building, I would guess it only does some parsing on the parameter, being considerably faster to build, is this assumption correct?

**Reply by ludralph** [reply]
> So there is a chance some of the more complex queries we have could be causing this, saturating the client side pool, but not translate to a high connection count in pgbouncer/db, is this correct?

Yes, it is possible that complex queries could be saturating the Prisma client-side pool without necessarily translating to a high connection count in PgBouncer or the database. This can happen if the queries are taking a long time to execute, causing the connections to be held longer and leading to timeouts.

> Some of our more complex queries involve deep nested select which sometimes require some extra filtering at the nested select level, do you think upgrading Prisma and trying the new join settings could help with this?

Upgrading your  Prisma version and utilizing new Join settings may optimize these complex queries, potentially reducing the load on the client-side pool. Improved Joins handling could translate to more efficient query execution and reduced connection usage.

> Any suggestions on how to identify potentially problematic queries other than adding prisma metrics to all suspect queries?

You can enable query logging at the database level to capture slow-running queries. Additionally, you can use tools like [Query Insights](https://cloud.google.com/sql/docs/postgres/using-query-insights) to monitor the performance of specific queries and see their CPU consumption:

> Also, I am assuming prisma Raw bypasses mostly the query building, I would guess it only does some parsing on the parameter, being considerably faster to build, is this assumption correct?

Using `prisma.$queryRaw`  bypasses most of the query building that Prisma does. It directly executes the raw SQL query, which can be considerably faster, especially for complex queries. This can help reduce the overhead associated with Prisma's query building and execution.

Also, I can suggest you try our new  [Driver Adapters](https://www.prisma.io/docs/orm/overview/databases/postgresql#using-the-node-postgres-driver) feature, which completely sidesteps our connection management, and relies on the native driver instead.

**Reply by joh9347** [reply]
@ludralph, I also want to ask a related question regarding this thread. Do you think the core issue could be the query engine build time instead of the long-running queries on the PostgreSQL side?

My thinking is that if the query engine builds the query quickly and sends it to PostgreSQL, we would still see that connection on the PostgreSQL side. Once the query request is sent to the database, Prisma should be waiting on that task and working on other query builds asynchronously.

If the query build time is extremely long and it doesn't even give a chance for other queries in the queue to be built, I think the metrics we are seeing here make sense. Is there a way we can track how long Prisma takes to build each query?

**Reply by ludralph** [reply]
Hi @joh9347

The query engine build time being a potential bottleneck is valid. If the query engine takes a long time to build the query before sending it to PostgreSQL, it could indeed cause delays and saturate the  Prisma client's connection pool, causing timeout errors without saturating PostgreSQL connections. This would prevent other queries from being built and executed promptly. We recommend identifying the slow-running queries and fetch only the required data. If you only need to use specific fields from the table, then you can use the select clause in your queries to prevent unnecessary data from being fetched. Additionally, you could consider adding appropriate indexes based on the query to ensure that data is fetched efficiently. You can also consider isolating long running queries into separate environments, so that these long running queries would not affect the transactional queries.

To track how long Prisma takes to build each query, you can enable [tracing](https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/opentelemetry-tracing). Tracing provides a detailed log of the activity that Prisma Client carries out, including the time taken to build and execute each query. This can help you identify if the query build time is contributing to the performance issues.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
